packages/EDA/src/softmax-classifier.ts

```diff
@@ -0,0 +1,390 @@
+
+import * as grok from 'datagrok-api/grok';
+import * as ui from 'datagrok-api/ui';
+import * as DG from 'datagrok-api/dg';
+
+import {_fitSoftmax} from '../wasm/EDAAPI';
+
+const ROWS_EXTRA = 1;
+const COLS_EXTRA = 2;
+const MIN_COLS_COUNT = 1 + COLS_EXTRA;
+const AVGS_NAME = 'Avg-s';
+const STDEVS_NAME = 'Stddev-s';
+const PRED_NAME = 'predicted';
+const DEFAULT_LEARNING_RATE = 1;
+const DEFAULT_ITER_COUNT = 100;
+const DEFAULT_PENALTY = 0.1;
+const DEFAULT_TOLERANCE = 0.001;
+const BYTES_PER_MODEL_SIZE = 4;
+
+/** Train data sizes */
+type DataSpecification = {
+  classesCount: number,
+  featuresCount: number,
+};
+
+/** Target labels specification */
+type TargetLabelsData = {
+  oneHot: Array<Uint8Array>,
+  weights: Uint32Array,
+};
+
+/** Softmax classifier */
+export class SoftmaxClassifier {
+  private avgs: Float32Array;
+  private stdevs: Float32Array;
+  private categories: string[];
+  private params: Float32Array[] | undefined = undefined;
+  private classesCount = 1;
+  private featuresCount = 1;
+
+  constructor(specification?: DataSpecification, packedModel?: Uint8Array) {
+    if (specification !== undefined) { // Create empty model
+      /** features count */
+      const n = specification.featuresCount;
+
+      /** classes count */
+      const c = specification.classesCount;
+
+      if (n < 1)
+        throw new Error('Incorrect features count');
+
+      if (c < 1)
+        throw new Error('Incorrect classes count');
+
+      /** length of arrays */
+      const len = n + ROWS_EXTRA;
+
+      // Init model routine
+      this.avgs = new Float32Array(len);
+      this.stdevs = new Float32Array(len);
+      this.categories = new Array<string>(len);
+      this.featuresCount = n;
+      this.classesCount = c;
+    } else if (packedModel !== undefined) { // Get classifier from packed model (bytes)
+      try {
+        // Extract model's bytes count
+        const sizeArr = new Uint32Array(packedModel.buffer, 0, 1);
+        const bytesCount = sizeArr[0];
+
+        // Model's bytes
+        const modelBytes = new Uint8Array(packedModel.buffer, BYTES_PER_MODEL_SIZE, bytesCount);
+
+        const modelDf = DG.DataFrame.fromByteArray(modelBytes);
+        const columns = modelDf.columns;
+        const colsCount = columns.length;
+
+        if (colsCount < MIN_COLS_COUNT)
+          throw new Error('incorrect columns count');
+
+        this.classesCount = colsCount - COLS_EXTRA;
+        this.featuresCount = modelDf.rowCount - ROWS_EXTRA;
+
+        const c = this.classesCount;
+
+        // extract params & categories
+        this.params = new Array<Float32Array>(c);
+        this.categories = new Array<string>(modelDf.rowCount);
+
+        for (let i = 0; i < c; ++i) {
+          const c = columns.byIndex(i);
+          this.categories[i] = c.name;
+
+          if (c.type !== DG.COLUMN_TYPE.FLOAT)
+            throw new Error('incorrect params columns type');
+
+          this.params[i] = c.getRawData() as Float32Array;
```

Comment:
Are we sure that the column is going to contains floats and not ints, for instance? Perhaps - I did not manage to trace it to the root